### PR TITLE
Create a V2 Docker plugin

### LINKED ISSUE #2397: work as a 'docker plugin' ('V2 plugin')
Docker introduces an experimental 'docker plugin ...' set of commands for managing plugins implemented as containers. We should make weave work in that way, though details are sketchy atm.

I imagine this issue will become a tracker for existing/new more specific issues that need to be resolved to pull this off.

### PR DESCRIPTION
**Pull request by bboreham**
Fixes #2397 #2816 #2651 #2805

Build the plugin via `make v2plugin`; requires Docker 1.13.1 (1.13.0 does not work)

This branch is based on top of the docker-swarm branch used in #2721 to get the extra entry-points, although it doesn't seem to make any difference.

Items outstanding:
- [x] [Connecting up peers](https://github.com/weaveworks/weave/issues/2805) - currently you have to `weave connect` to test
- [x] Docker don't support `--pid=host` on v2 plugins; need another way to do the arp settings on the bridge
- [x] No access to the Docker socket inside the plugin, so the hack to do arp settings on the veth doesn't work
- [x] No access to the Docker socket inside the plugin, so need to persist anything we are told on network creation (e.g. the multicast option)
- [x] `weave status` does work for the router, but doesn't see the plugin
- [x] `weave launch` should detect if this plugin is installed and exit with error
- [x] Better UX for troubleshooting - can't do `docker logs` to see what went wrong

### CODE REVIEW

**Comment by sgammill** [comment]
@bboreham Hello, can you give an update on the status of weave net in docker swarm-mode since https://github.com/docker/swarmkit/pull/1876 has happened and docker engine is now at 1.13.1-rc1?  Thanks.

**Comment by brb** [comment]
@sgammill I verified that the blocking issue https://github.com/docker/docker/issues/30024 got resolved, and Weave Net Plugin v2 works with Docker Swarmkit (distributed with 1.31.1-rc1).

However, to make it usable to an end user, more work needs to be done (see the outstanding items in the PR description).

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by marccarre** [comment]
- `shfmt`-ed  `launch.sh` to please `make lint` in CircleCI.
- Added `DOCKERHUB_USER` to plugin's name (more convenient for development, and prob. what we want ultimately?)
- Bind-mounted `/proc`, which should help passing ARP settings:

      "mounts": [
        {
          "destination": "/procmount",
          "type": "proc",
          "source": "proc"
        }
      ],

    Manual testing:
    
      $ sudo docker-runc exec 512af90f342f16dddebc7b7245985d5eb67557e76798b4033eed2ff7adaa4eb5 cat /proc/sys/net/ipv4/neigh/weave/ucast_solicit
      3
      $ cat /proc/sys/net/ipv4/neigh/weave/ucast_solicit
      3
      # echo 1 > /proc/sys/net/ipv4/neigh/weave/ucast_solicit
      $ cat /proc/sys/net/ipv4/neigh/weave/ucast_solicit
      1
      $ sudo docker-runc exec 512af90f342f16dddebc7b7245985d5eb67557e76798b4033eed2ff7adaa4eb5 cat /proc/sys/net/ipv4/neigh/weave/ucast_solicit 
      1

- Tried to bind-mount `/var/lib/weave` and `/var/log/weave` in various ways, in order to respectively persist data and logs, in a way which makes them accessible to the host (reload data if container gets destroyed, ease of debugging, etc.), but faced issues. I'm probably missing something obvious.

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by marccarre** [comment]
Note that, given:

- we use Docker 1.13.1 in this branch, and 
- because of docker/docker/issues/31447, 

tests 335 and 620 currently fail (time out).

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

File: weave [hunk 1]
```diff
@@ -883,6 +883,24 @@ attach() {
...
+}
+
+# Docker Plugin V2
+is_docker_plugin_enabled() {
+    util_op is-docker-plugin-enabled $PLUGIN_V2_NAME 2>/dev/null
+}
+
+# Docker Plugin (part of weaver process, so V2 or V1).
+# Assumes that plugin is running inside Docker container.
+is_plugin_enabled() {
```
Comment (marccarre) [inline weave]: Should we actually rename these two functions so that it is clearer which one applies to which plugin?
e.g.:
```
is_plugin_enabled        -> is_plugin_v1_enabled
is_docker_plugin_enabled -> is_plugin_v2_enabled
```

File: api/weaveapi.go [hunk 2]
```diff
@@ -76,6 +77,22 @@ func (client *Client) Connect(remote string) error {
...
+// IsReady returns true if the API server is up and running
+func (client *Client) IsReady() bool {
+	_, err := client.httpVerb("GET", "/status", nil)
+
+	return err == nil
+}
+
+// WaitAPIServer waits until the API server is ready to serve.
+//
+// Might block infinitely.
```
Comment (marccarre) [inline api/weaveapi.go]: Questions:
1. Do we actually want to retry indefinitely? 
2. Will failures be obvious when debugging and/or user support happens?
3. Is there any other information we could leverage to "fail faster" or, alternatively, should we give up after some amount of time?

**Comment by awh** [comment]
#### Documentation changes:
modified [plugin](https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/)

**Comment by panga** [comment]
Where I can find a working version to use with Docker Swarm 17.03+?

**Comment by bboreham** [comment]
@panga we haven't released one; feel free to build from source on this branch and tell us how you get on.

**Comment by panga** [comment]
@bboreham Thank you, I'll try from source.

I can't access the updated docs, this link is correctly?
https://dev-weavewww.pantheonsite.io/docs/net/2397-v2-plugin/plugin/

File: site/plugin.md [hunk 3]
```diff
@@ -59,8 +59,29 @@ then run your containers using the Docker command-line:
...
-["key-value store"](https://docs.docker.com/engine/userguide/networking/get-started-overlay/#step-1-set-up-a-key-value-store).
+this between hosts requires that you run Docker in ["swarm mode"](https://docs.docker.com/engine/swarm/swarm-mode/) or configure a
+["key-value store"](https://docs.docker.com/engine/userguide/networking/get-started-overlay/#/set-up-a-key-value-store).
+
+Docker swarm mode requires Docker version 1.13 or later to work with
+plugins such as Weave Net.
+
+To create a new network for services in swarm mode, run:
+
+    $ docker network create --driver=weave mynetwork
```
Comment (bboreham) [inline site/plugin.md]: don't we need to give the DockerHub url of the plugin here for V2?

**Comment by bboreham** [comment]
@panga That is our preview version of the main WeaveWorks website; it's not open to the world.  But you can read the Weave Net docs from this PR [on github](https://github.com/weaveworks/weave/blob/5f56242ed905ae81e037369b281c628453967dda/site/plugin.md)

File: site/plugin.md (hunk 3 above)
Comment (brb) [inline site/plugin.md]: The docs update is not part of this PR

File: weave (hunk 1 above)
Comment (brb) [inline weave]: Done.

File: weave [hunk 4]
```diff
@@ -2062,7 +2055,11 @@ EOF
             echo "        Service: proxy"
             echo "        Address: $PROXY_ADDRS"
         fi
-        if [ -z "$SUB_STATUS" ] && check_running $PLUGIN_CONTAINER_NAME 2>/dev/null ; then
+        if [ -z "$SUB_STATUS" ] && is_docker_plugin_enabled ; then
+            echo
+            echo "        Service: plugin"
```
Comment (brb) [inline weave]: Done. `Service: plugin (v1)` and `Service: plugin (v2)`.

File: prog/weaveutil/plugin_network.go [hunk 5]
```diff
@@ -71,6 +73,34 @@ while Weave is not running`, networkName, containers)
...
+
+	// This is messed up: we are using docker/docker/client instead of
+	// fsouza/go-dockerclient because the latter does not support plugins.
+	c, err := client.NewEnvClient()
+	if err != nil {
+		return fmt.Errorf("unable to connect to docker: %s", err)
+	}
+
+	ctx := context.Background()
+	p, _, err := c.PluginInspectWithRaw(ctx, pluginName)
```
Comment (brb) [inline prog/weaveutil/plugin_network.go]: the cmd which calls the noted function is invoked only from `is_plugin_v2_enabled` in the weave script. so, in the case of the older Docker version, the noted function should fail, and no plugin-v2 will be detected. this is fine as the older version does not support v2 plugins anyway.

Comment (brb) [inline prog/weaveutil/plugin_network.go]: I will add a comment to the cmd.

File: plugin/plugin.go [hunk 6]
```diff
@@ -21,47 +18,29 @@ import (
...
-	flag.Parse()
-
-	if justVersion {
-		fmt.Printf("weave plugin %s\n", version)
-		os.Exit(0)
-	}
-
-	common.SetLogLevel(logLevel)
+func Start(weaveAPIAddr string, dockerAPIAddr string, address string, meshAddress string) {
+	weave := weaveapi.NewClient(weaveAPIAddr, Log)
```
Comment (brb) [inline plugin/plugin.go]: It is in my TODO. For this PR I didn't want to introduce any other significant changes to prevent from possible bugs.

File: plugin/plugin.go [hunk 7]
```diff
@@ -21,47 +18,29 @@ import (
...
-
-	// API 1.21 is the first version that supports docker network commands
-	dockerClient, err := docker.NewVersionedClientFromEnv("1.21")
-	if err != nil {
-		Log.Fatalf("unable to connect to docker: %s", err)
+	var dockerClient *docker.Client
+	var err error
+	if dockerAPIAddr != "" {
+		// API 1.21 is the first version that supports docker network commands
+		dockerClient, err = docker.NewVersionedClient(dockerAPIAddr, "1.21")
```
Comment (brb) [inline plugin/plugin.go]: Due to different versions. Anyway, I made it re-use the one from weaver which uses unversioned instance of Docker client. After inspecting `go-dockerclient`, I don't see any point of pinning versions (besides min-api-vsn check in dockerd which is not relevant to us).

File: prog/weaver/main.go [hunk 8]
```diff
@@ -368,6 +376,10 @@ func main() {
 		go listenAndServeHTTP(statusAddr, statusMux)
 	}
 
+	if enablePlugin {
+		go plugin.Start(httpAddr, dockerAPI, pluginSocket, pluginMeshSocket)
```
Comment (brb) [inline prog/weaver/main.go]: When initializing `go-dockerclient`, the initialization routine `NewClient` sends the `/version` request. If Docker is not listening on the socket, the request will fail, the weaver process will crash (due to `log.Fatal`), and weaver will get restarted by Docker.

File: plugin/net/driver.go [hunk 9]
```diff
@@ -169,6 +187,9 @@ func (driver *driver) findNetworkInfo(id string) (network, error) {
 	if found {
 		return network, nil
 	}
+	if driver.docker == nil { // No way to look it up - TODO persist the info
```
Comment (brb) [inline plugin/net/driver.go]: Thanks for noting. This helped to find a bug in the plugin-v2 integration code.

File: weave [hunk 10]
```diff
@@ -2010,17 +2010,10 @@ EOF
...
         ;;
-    launch-plugin)
-        if ! check_running $CONTAINER_NAME 2>/dev/null ; then
-            echo "ERROR:" $CONTAINER_NAME "container must be running before plugin can be launched" >&2
-            exit 1
-        fi
-        launch_plugin_if_not_running "$@"
-        echo $PLUGIN_CONTAINER
-        ;;
     env|proxy-env)
```
Comment (brb) [inline weave]: Are we going to print some message in a case of `weave run`? If yes, then I should probably print some warning as well.

File: weave [hunk 11]
```diff
@@ -2251,7 +2248,6 @@ EOF
         ;;
     stop)
         [ $# -eq 0 ] || usage
-        plugin_disabled || stop_plugin
```
Comment (brb) [inline weave]: I was thinking about giving explicit instructions in release notes instead. But anyway, I don't have a strong opinion about it though.

File: weave [hunk 12]
```diff
@@ -2287,7 +2279,7 @@ EOF
                 exit 1
                 ;;
         esac
-        for NAME in $PLUGIN_CONTAINER_NAME $CONTAINER_NAME $PROXY_CONTAINER_NAME ; do
```
Comment (brb) [inline weave]: https://github.com/weaveworks/weave/pull/2727#discussion_r110210039

File: api/weaveapi.go (hunk 2 above)
Comment (brb) [inline api/weaveapi.go]: I've changed to waiting for 1min. Also, logging is explicit.

**Comment by brb** [comment]
PTAL.

File: plugin/net/watcher.go [hunk 13]
```diff
@@ -41,11 +41,19 @@ func (w *watcher) ContainerStarted(id string) {
 			continue
 		}
 		if network.isOurs {
-			fqdn := fmt.Sprintf("%s.%s", info.Config.Hostname, info.Config.Domainname)
-			if err := w.weave.RegisterWithDNS(id, fqdn, net.IPAddress); err != nil {
-				w.driver.warn("ContainerStarted", "unable to register %s with weaveDNS: %s", id, err)
+			if !w.driver.noDNS {
```
Comment (marccarre) [inline plugin/net/watcher.go]: *Minor*: double negation. Suggestion: `dnsEnabled bool` instead of `noDNS bool`.

File: plugin/plugin.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+		return err
+	}
+	go serveStatus(statusListener)
+
+	return <-endChan
+}
+
+func listenAndServe(dockerClient *docker.Client, weave *weaveapi.Client, address string, endChan chan<- error, scope string, withIpam, noDNS bool, isPluginV2 bool) (net.Listener, error) {
+	name := strings.TrimSuffix(path.Base(address), ".sock")
+	// TODO(mp) fix plugin name, as it is used to determine whether network is our
```
Comment (marccarre) [inline plugin/plugin.go]: Was the intent to fix this `TODO` before the PR gets merged?

Comment (brb) [inline plugin/plugin.go]: I'm going to fix it.

File: plugin/net/watcher.go (hunk 13 above)
Comment (brb) [inline plugin/net/watcher.go]: Addressed.

File: plugin/plugin.go (hunk 14 above)
Comment (brb) [inline plugin/plugin.go]: Addressed, but in an ugly way:(

The problem is that Docker Plugin V2 does not expose a plugin name, so we need to search for `"net-plugin"` within a driver name of a network. Anyway, suggestions welcome.

File: prog/weaver/main.go [hunk 15]
```diff
@@ -407,7 +407,7 @@ func main() {
 	}
 
 	if enablePlugin || enablePluginV2 {
-		go plugin.Start(httpAddr, dockerCli, pluginSocket, pluginMeshSocket, noDNS, enablePluginV2)
+		go plugin.Start(httpAddr, dockerCli, pluginSocket, pluginMeshSocket, !noDNS, enablePluginV2)
```
Comment (marccarre) [inline prog/weaver/main.go]: `!noDNS` is most likely an omission, I'll rename it later.

Comment (marccarre) [inline prog/weaver/main.go]: Actually, this is valid.
It matches the CLI flag `-no-dns`.
It is then negated since the rest of the code cares about `dns`.

**Comment by bboreham** [comment]
@panga note you can now install our "latest" build from DockerHub, via `docker plugin install weaveworks/net-plugin`

**Comment by Multiply** [comment]
@bboreham Any documentation publicly available?

**Comment by bboreham** [comment]
@Multiply we're working on it.  If your swarm is up and running, and you install as above, next step is:

    docker network create --driver=weaveworks/net-plugin mynetworkname

then

    docker run --net=mynetworkname ...

everything else is automagic 🙂

**Comment by Multiply** [comment]
@bboreham Sweet. I'll give it a spin over the weekend. I think it'll be more or less plug and play with our current development and testing setups.

**Comment by bboreham** [comment]
oh, wait, you have to say `--attachable` when you create the network if you want to do `docker run`

We need to finish the docs

**Comment by Multiply** [comment]
I should be able to define it using docker-compose v2/v3 and have it create the needed network plugin for us, if it doesn't already exist, as long as they installed the plugin beforehand, I assume.

Would be even sweeter if it could also automatically install the plugin if its not available. (I haven't looked much into the documentation, so it might already be like this?)

**Comment by bboreham** [comment]
One thing is, when the plugin installs, Docker asks you whether you want to grant the necessary permissions (which is a good thing for knowledgeable users) - we need `NET_ADMIN` and stuff like that.  So I don't know how this would work in an automatic install situation.

Looking at [the docs](https://docs.docker.com/engine/reference/commandline/plugin_install/) there is a `--grant-all-permissions` flag.

Nothing [in the compose docs](https://docs.docker.com/compose/networking/#specifying-custom-networks) about installing plugins.

**Comment by jamespedwards42** [comment]
@bboreham I was able to install the plugin, but when trying to create a network I get the following error:
```
> docker plugin ls
ID                  NAME                           DESCRIPTION                   ENABLED
667779c5ab70        weaveworks/net-plugin:latest   Weave Net plugin for Docker   true
81bd50710a37        rexray/ebs:latest              REX-Ray for Amazon EBS        true
```

```
> docker network create --driver=weaveworks/net-plugin --attachable mynetworkname
Error response from daemon: could not resolve driver weaveworks/net-plugin in registry
```

```
> docker version
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by brb** [comment]
@jamespedwards42 you should pass the whole name of the driver when creating a network, i.e.: `docker network create --driver=weaveworks/net-plugin:latest --attachable mynetworkname`

**Comment by jamespedwards42** [comment]
@brb Same isssue:
```
> docker network create --driver=weaveworks/net-plugin:latest --attachable mynetworkname
Error response from daemon: could not resolve driver weaveworks/net-plugin:latest in registry
```

**Comment by brb** [comment]
@jamespedwards42 Could you do the following:

* Check whether `weaver` process is running (`ps aux | grep weaver`).
* Provide docker logs (`journalctl -r -u docker.service` if you run systemd).

**Comment by jamespedwards42** [comment]
@brb I restarted the docker service and then ran the create network command for the logs capture; pasted on pastebin.

```
> ps aux | grep weaver
root      5068  0.0  0.0  12944   960 pts/0    S+   13:13   0:00 grep --color=auto weaver
```

`> journalctl -r -u docker.service`
https://pastebin.com/sxjr0kyC

**Comment by brb** [comment]
That's weird.

Mind running `docker plugin disable weaveworks/net-plugin:latest && docker plugin enable weaveworks/net-plugin:latest`? If the problem persists, please provide the `docker.service` logs again. Thanks.

**Comment by jamespedwards42** [comment]
@brb ha! awesome that got it.... I restarted a server and the issue came back, so I had to disable/enable again.  I'm guessing that needs to be done anytime the engine is restarted.

@Multiply I was able to confirm that it works with docker stack deploy against a docker compose file.  It created the configured network for me:

```yml
networks:
  my-weave-network:
    driver: weaveworks/net-plugin:latest
    ipam:
      driver: default
      config:
        -
          subnet: 10.0.0.0/24
```

**Comment by brb** [comment]
> I'm guessing that needs to be done anytime the engine is restarted.

I've just tried rebooting machine / restarting the engine - the plugin came back properly.

**Comment by jamespedwards42** [comment]
@brb I just went through 4 hosts and had the same issue after restarts on all of them.

```
> docker network create --driver=weaveworks/net-plugin:latest --attachable mynetworkname
Error response from daemon: could not resolve driver weaveworks/net-plugin:latest in registry

> docker plugin disable weaveworks/net-plugin:latest weaveworks/net-plugin:latest

> docker plugin enable weaveworks/net-plugin:latest weaveworks/net-plugin:latest

> docker network create --driver=weaveworks/net-plugin:latest --attachable mynetworkname
mc9pap4xanv1f94cuxdmsgek7
```

logs after restart: https://pastebin.com/B4ngjtnt

**Comment by brb** [comment]
@jamespedwards42 Thanks. Found the issue (https://github.com/weaveworks/weave/issues/2906).
